src/Microsoft.AspNet.SignalR.Client/Connection.cs

```diff
@@ -469,39 +473,80 @@ protected virtual string OnSending()
+                // If we have a keep alive
+                if (negotiationResponse.KeepAliveTimeout != null)
+                {
+                    _keepAliveData = new KeepAliveData(TimeSpan.FromSeconds(negotiationResponse.KeepAliveTimeout.Value));
+                    _reconnectWindow = _disconnectTimeout + _keepAliveData.Timeout;
 
-                                // If we have a keep alive
-                                if (negotiationResponse.KeepAliveTimeout != null)
-                                {
-                                    _keepAliveData = new KeepAliveData(TimeSpan.FromSeconds(negotiationResponse.KeepAliveTimeout.Value));
-                                    _reconnectWindow = _disconnectTimeout + _keepAliveData.Timeout;
+                    beatInterval = _keepAliveData.CheckInterval;
+                }
+                else
+                {
+                    _reconnectWindow = _disconnectTimeout;
+                }
+
+                Monitor = new HeartbeatMonitor(this, _stateLock, beatInterval);
+
+                return StartTransport();
+            }
 
-                                    beatInterval = _keepAliveData.CheckInterval;
-                                }
-                                else
+            Task StartNegotiation()
+            {
+                return transport.Negotiate(this, _connectionData)
+                                .Then(negotiationResponse =>
                                 {
-                                    _reconnectWindow = _disconnectTimeout;
-                                }
+                                    VerifyProtocolVersion(negotiationResponse.ProtocolVersion);
+
+                                    if (negotiationResponse.ProtocolVersion.Equals("2.0") && !string.IsNullOrEmpty(negotiationResponse.RedirectUrl))
+                                    {
+                                        // Ensure the URL ends with a trailing slash.
+                                        if (!negotiationResponse.RedirectUrl.EndsWith("/"))
+                                        {
+                                            negotiationResponse.RedirectUrl += "/";
+                                        }
+
+                                        // Update the URL based on the redirect response and restart the negotiation
+                                        Url = negotiationResponse.RedirectUrl.ToString();
+
+                                        if (!string.IsNullOrEmpty(negotiationResponse.AccessToken))
+                                        {
+                                            // This will stomp on the current Authorization header, but that's by design.
+                                            // If the server specified a token, that is expected to overrule the token the client is currently using.
+                                            Headers["Authorization"] = $"Bearer {negotiationResponse.AccessToken}";
```

Comment:
The `AccessToken` should also be applied to other requests `/reconnect`, `/ping`, `/send`.. etc. And for `ws` and `sse` requests, it should be appended in the query string.

Reply:
It doesn't need to be in the query string for any client but the javascript client.

Reply:
Yep, we don't have the same limitations in .NET that we have in the browser. It's only the browser that forbids headers in EventSource and WebSocket requests.

Reply:
Knowledge get!